File: python/tradingflow/portfolio/mean_variance/admm_mnr.py
"""Mean-variance optimization by matrix-free ADMM-MNR — NumPy and SciPy only.

Solves the same problem as `markowitz` in `MIN_MEAN_VARIANCE` mode, maximizing
`muᵀx - delta xᵀΣx` subject to a budget and a box, but without CVXPY. The
covariance is never formed or factorized inside the solve: it enters only as
the operator `x -> B(Bᵀx) + d²x`, so each iteration costs `O(N k)` rather than
`O(N²)`, and there is no canonicalization step to amortize.
"""

import numpy as np

from .._base import MeanVariancePortfolio
from .._factor import factor_decompose
from . import _admm


def solve(
    mu: np.ndarray,
    sigma: np.ndarray,
    *,
    delta: float,
    long_only: bool,
    cap: float | None,
    factor_rank: int | None,
    outer_tol: float,
    max_outer: int,
) -> np.ndarray:
    """Decomposes the covariance and runs the ADMM-MNR solve."""
    n = sigma.shape[0]

    if factor_rank is None:
        # Dense path: the solver builds its own operator from an eigen split of
        # the covariance. Exact, but O(N^3) per rebalance.
        result = _admm.solve_portfolio(
            covariance=sigma,
            expected_returns=mu,
            delta=delta,
            lower=0.0 if long_only else -np.inf,
            cap=cap,
            outer_tol=outer_tol,
            max_outer=max_outer,
        )
    else:
        loadings, idio = factor_decompose(sigma, factor_rank)
        result = _admm.solve_portfolio(
            factor_loadings=loadings,
            specific_var=idio**2,
            expected_returns=mu,
            delta=delta,
            lower=0.0 if long_only else -np.inf,
            cap=cap,
            outer_tol=outer_tol,
            max_outer=max_outer,
        )

    weights = np.asarray(result["weights"], dtype=np.float64)
    if not np.all(np.isfinite(weights)):
        return np.full(n, 1.0 / n)
    return np.maximum(weights, 0.0) if long_only else weights


def build(
    *,
    delta: float,
    long_only: bool = True,
    cap: float | None = None,
    factor_rank: int | None = 20,
    outer_tol: float = 1e-3,
    max_outer: int = 400,
    **kwargs,
) -> MeanVariancePortfolio:
    """Constructs a matrix-free ADMM mean-variance portfolio.

    `delta` is the risk-aversion coefficient, `cap` an optional per-name upper
    bound, and `factor_rank` the rank of the covariance approximation the
    matrix-free operator is built from — `None` switches to the exact but cubic
    dense path.
    """
    assert delta > 0.0, "delta must be positive"
    return MeanVariancePortfolio(
        solve=lambda state, active, universe, previous, mu, sigma: solve(
            mu,
            sigma,
            delta=float(delta),
            long_only=long_only,
            cap=cap,
            factor_rank=factor_rank,
            outer_tol=outer_tol,
            max_outer=max_outer,
        ),
        **kwargs,
    )

File: crates/tradingflow/src/operators/portfolio/mean_variance/admm_mnr.rs
use pyo3::types::PyDictMethods;

use super::MeanVariancePortfolio;
use crate::operators::portfolio::Config;
use crate::python::py_segment_module;

/// Mean-variance optimization by matrix-free ADMM, without CVXPY.
///
/// Solves the same problem as [`markowitz`](super::markowitz) in
/// [`Mode::MinMeanVariance`](super::Mode::MinMeanVariance) — maximize
/// `μᵀx - delta · xᵀΣx` subject to a budget and a box — but reaches it through
/// an ADMM outer loop over MINRES solves of the saddle system, so the
/// covariance enters only as the operator `x ↦ B(Bᵀx) + d²x`. Nothing is ever
/// factorized inside the solve, each iteration costs `O(N k)` rather than
/// `O(N²)`, and there is no canonicalization to amortize.
///
/// The practical difference from `markowitz`: no CVXPY dependency, so this runs
/// wherever NumPy and SciPy do, and cost scales with the factor rank rather
/// than the cross-section. The cost is an iterative solve with tolerances to
/// tune rather than a solver that either converges or reports why not.
///
/// `delta` is the risk-aversion coefficient, `cap` an optional per-name upper
/// bound, and `factor_rank` the rank of the covariance approximation the
/// operator is built from — `None` uses the exact covariance, at `O(N³)` per
/// rebalance.
///
/// See [module-level docs](super::super) for inputs and outputs.
///
/// # Panics
///
/// If `delta` is not positive.
pub fn admm_mnr(
    config: Config,
    delta: f64,
    long_only: bool,
    cap: Option<f64>,
    factor_rank: Option<usize>,
) -> impl MeanVariancePortfolio {
    assert!(delta > 0.0, "admm_mnr delta must be positive, got {delta}");
    py_segment_module(
        "tradingflow.portfolio.mean_variance.admm_mnr",
        config.params(|d| {
            d.set_item("delta", delta)?;
            d.set_item("long_only", long_only)?;
            d.set_item("cap", cap)?;
            d.set_item("factor_rank", factor_rank)
        }),
    )
}
